File: Android/doraemonkit/src/main/java/com/didichuxing/doraemonkit/widget/tableview/utils/DensityUtils.java
package com.didichuxing.doraemonkit.widget.tableview.utils;

import android.content.Context;
import android.util.TypedValue;



public class DensityUtils {

    private DensityUtils() {
    }


    /**
     * dp转px
     *
     * @param context
     * @return
     */

    public static int dp2px(Context context, float dpVal) {

        return (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP,

                dpVal, context.getResources().getDisplayMetrics());

    }


    /**
     * sp转px
     *
     * @param context
     * @return
     */

    public static int sp2px(Context context, float spVal) {

        return (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP,

                spVal, context.getResources().getDisplayMetrics());

    }


    /**
     * px转dp
     *
     * @param context
     * @param pxVal
     * @return
     */

    public static float px2dp(Context context, float pxVal) {

        final float scale = context.getResources().getDisplayMetrics().density;

        return (pxVal / scale);

    }


    /**
     * px转sp
     *
     * @param pxVal
     * @return
     */

    public static float px2sp(Context context, float pxVal) {

        return (pxVal / context.getResources().getDisplayMetrics().scaledDensity);

    }


}

File: Android/dokit-mc/src/main/java/com/didichuxing/doraemonkit/kit/mc/utils/DensityUtils.kt
package com.didichuxing.doraemonkit.kit.mc.utils

import android.content.Context

/**
 * ================================================
 * 作    者：jint（金台）
 * 版    本：1.0
 * 创建日期：2020/12/10-17:00
 * 描    述：
 * 修订历史：
 * ================================================
 */
/**
 * px与dp互相转换
 * Created by yangle on 2016/4/12.
 */
object DensityUtils {
    fun dp2px(context: Context, dp: Float): Int {
        //获取设备密度
        val density = context.resources.displayMetrics.density
        //4.3, 4.9, 加0.5是为了四舍五入
        return (dp * density + 0.5f).toInt()
    }

    fun px2dp(context: Context, px: Int): Float {
        //获取设备密度
        val density = context.resources.displayMetrics.density
        return px / density
    }
}
